### PR TITLE
[bugfix] Resolve memory not logged when missing metrics

### LINKED ISSUE #8159: Cannot log gpu memory without making on step = True while logging train loss
### Discussed in https://github.com/PyTorchLightning/pytorch-lightning/discussions/7582

<div type='discussions-op-text'>

<sup>Originally posted by **karthi0804** May 18, 2021</sup>
I am using 1.3.0 version. I wanted to just log the gpu memory at all time steps. So I enabled ```log_gpu_memory='all'``` and ```log_every_n_steps=1```. I couldn't find the gpu logging for steps. But If I log the train loss on steps, I can see the gpu logging at each steps. Is this how it is intended to work? I expect that gpu logging should be independent of self.log(). </div>

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

This PR adds gpus_metrics to ResultCollection + filter the non requested gpus for logging

Fixes #8159

## Before submitting
- [ ] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [ ] Did you make sure to update the documentation with your changes? (if necessary)
- [ ] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py [hunk 1]
```diff
@@ -226,6 +224,12 @@ def update_train_epoch_metrics(self) -> None:
         # reset result collection for next epoch
         self.trainer._results.reset(metrics=True)
 
+    def _log_gpus_metrics(self):
+        for key, mem in self.gpus_metrics.items():
+            gpu_id = int(key.split('/')[0].split(':')[1])
+            if gpu_id in self.trainer.accelerator_connector.parallel_device_ids:
```
Comment (justusschock) [inline pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py:230]: Is this also set for just a single gpu?

**Review by justusschock — APPROVED** [review]

**Review by ethanwharris — APPROVED** [review]
LGTM :smiley:

File: pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py (hunk 1 above)
Comment (tchaton) [reply pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py:230]: Screenshot 2021-06-28 at 14 33 32

Yes, it works.

**Review by awaelchli — APPROVED** [review]

File: pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py [hunk 2]
```diff
@@ -226,6 +224,12 @@ def update_train_epoch_metrics(self) -> None:
...
+                self.trainer.lightning_module.log(key, mem, prog_bar=False, logger=True, on_step=True, on_epoch=False)
```
Comment (ananthsub) [inline pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py:231]: since we're already in the trainer, why do we have to log through the lightning module's `log` ?

File: pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py [hunk 3]
```diff
@@ -276,6 +280,13 @@ def metrics(self) -> Dict[MetricSource, Dict[str, _METRIC]]:
         on_step = not self._epoch_end_reached
         return self.trainer._results.metrics(on_step)
 
+    @property
+    def gpus_metrics(self) -> Dict[str, str]:
+        if self.trainer._device_type == DeviceType.GPU and self.log_gpu_memory:
+            mem_map = memory.get_memory_profile(self.log_gpu_memory)
+            self._gpus_metrics.update(mem_map)
+        return self._gpus_metrics
+
```
Comment (carmocca) [inline pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py:289]: Why did this PR need to add `_gpu_metrics`? Doesn't seem related to the issue linked.

This means that gpu metrics are now duplicated in this dictionary and in logged metrics.

Also it only gets filled when `self.log_gpu_memory` so it can't be used anyways without the flag.

Comment (thomasw21) [reply pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py:289]: I think this broke `log_gpu_memory="min_max"` option in `Trainer.fit`. Looks related to: https://github.com/PyTorchLightning/pytorch-lightning/issues/9010

Essentially `memory. get_memory_profile` adds two keys that are not in conventional format `min_gpu_mem` and `max_gpu_mem` (typically keys are in `f"gpu_id: {gpu_id}/memory.used (MB)"`
